app/gui/src/model/suggestion_database.rs

```diff
@@ -27,6 +29,108 @@ pub use example::Example;
+        let value = Some(id);
+        let old_value = self.replace_value_and_traverse_back_pruning_empty_subtrees(path, value);
+        if old_value.is_some() {
+            event!(WARN, "An existing suggestion entry id at {path:?} was overwritten with {id}.");
+        }
+    }
+
+    /// Removes the [`entry::Id`] stored at `path`. Emits a warning if there was no [`entry::Id`]
+    /// stored at `path` before the operation.
+    pub fn warn_if_absent_and_remove(&mut self, path: &entry::QualifiedName) {
+        let old_value = self.replace_value_and_traverse_back_pruning_empty_subtrees(path, None);
+        if old_value.is_none() {
+            let msg = format!(
+                "Could not remove a suggestion entry id at {path:?} because it does not exist."
+            );
+            event!(WARN, "{msg}");
+        }
+    }
+
+    /// Sets the value at `path` to `value` and returns the replaced value (returns `None` if there
+    /// was no node at `path`). Then visits nodes on the `path` in reverse order and removes every
+    /// visited empty leaf node from its parent.
+    ///
+    /// A node is defined as empty when it contains a `None` value. A node is a leaf when its
+    /// [`is_leaf`] method returns `true`.
+    ///
+    /// The function is optimized to not create new empty nodes if they would be deleted by the
+    /// function before it returns.
+    fn replace_value_and_traverse_back_pruning_empty_subtrees(
+        &mut self,
+        path: &entry::QualifiedName,
+        value: Option<entry::Id>,
+    ) -> Option<entry::Id> {
+        // A helper of the [`replace_value_and_traverse_back_pruning_empty_subtrees`] function.
+        // Performs the same operation but the replaced value is swapped with `value`
+        // instead of being returned.
+        fn swap_value_and_traverse_back_pruning_empty_subtrees<P, I>(
+            node: &mut HashMapTree<entry::QualifiedNameSegment, Option<entry::Id>>,
+            mut path: P,
+            value: &mut Option<entry::Id>,
+        ) where
+            P: Iterator<Item = I>,
+            I: Into<entry::QualifiedNameSegment>,
+        {
+            use std::collections::hash_map::Entry;
+            match path.next() {
+                None => std::mem::swap(&mut node.value, value),
+                Some(key) => match node.branches.entry(key.into()) {
+                    Entry::Occupied(mut entry) => {
+                        let node = entry.get_mut();
+                        swap_value_and_traverse_back_pruning_empty_subtrees(node, path, value);
+                        if node.value.is_none() && node.is_leaf() {
+                            entry.remove_entry();
+                        }
+                    }
+                    Entry::Vacant(entry) =>
+                        if let Some(v) = value.take() {
+                            entry.insert(default()).set(path, Some(v));
+                        },
+                },
+            }
+        }
```

Comment:
Why are we making the function `replace_value_and_traverse_back_pruning_empty_subtrees` artificially so long? Does this function really need to be defined inside of the other function? What are the benefits of defining functions this way vs defining them side by side?